webapp/src/components/ProgramActivityDialog.tsx

```diff
@@ -1,85 +1,212 @@
 import {
-  Button,
+  Alert,
   Dialog,
   DialogActions,
   DialogContent,
   DialogContentText,
   DialogTitle,
   Divider,
+  FormGroup,
+  Snackbar,
   Table,
   TableBody,
   TableCell,
   TableContainer,
   TableRow,
 } from '@mui/material';
-
+import IconButton from '@mui/material/IconButton';
 import { formatDate, formatTime } from '../helpers/dateTime';
 import { CalendarEvent } from '../types/api';
-
 interface ProgramActivityDialogProps {
   show: boolean;
   contents: CalendarEvent;
   handleClose: () => void;
 }
+const sendAPIGetRequest = (
+  path: string,
+  activityType: string,
+  setCompleted: React.Dispatch<React.SetStateAction<boolean | null>>,
+  setError: React.Dispatch<React.SetStateAction<boolean | null>>,
+  setIsErrorShown: React.Dispatch<React.SetStateAction<boolean>>,
+  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>
+) => {
+  setIsLoading(true);
+  if (activityType !== 'assignment') return;
+  return fetch(`${process.env.REACT_APP_API_ORIGIN}${path}`, {
+    method: 'GET',
+    credentials: 'include',
+  })
+    .then(res => res.json())
+    .then(
+      async ({ data }) => {
+        if (data) {
+          setIsLoading(false);
+          setError(false);
+          setCompleted(data.completed);
+        }
+      },
+      error => {
+        setIsLoading(false);
+        setError(true);
+        setIsErrorShown(true);
+      }
+    );
+};
+
+const sendAPIPutRequest = (
+  path: string,
+  body: Object,
+  setCompleted: React.Dispatch<React.SetStateAction<boolean | null>>,
+  setIsUpdateSuccess: React.Dispatch<React.SetStateAction<boolean>>,
+  setIsMessageVisible: React.Dispatch<React.SetStateAction<boolean>>,
+  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>
+) => {
+  setIsLoading(true);
+  if ('completed' in body && body.completed === null) {
+    return;
+  }
+
+  fetch(`${process.env.REACT_APP_API_ORIGIN}${path}`, {
+    method: 'PUT',
+    credentials: 'include',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify(body),
+  })
+    .then(res => res.json())
+    .then(({ data }) => {
+      if (data) {
+        setIsLoading(false);
+        setCompleted(data.completed);
+        setIsUpdateSuccess(true);
+        setIsMessageVisible(true);
+      }
+    })
+    .catch(error => {
+      setIsLoading(false);
+      setIsUpdateSuccess(false);
+      setIsMessageVisible(true);
+    });
+};
 
 const ProgramActivityDialog = ({
   show,
   contents,
   handleClose,
 }: ProgramActivityDialogProps) => {
+  const [completed, setCompleted] = useState<boolean | null>(null);
+  const [isUpdateSuccess, setIsUpdateSuccess] = useState<boolean>(false);
+  const [isMessageVisible, setIsMessageVisible] = useState<boolean>(false);
+  const [error, setError] = useState<boolean | null>(null);
+  const [isErrorShown, setIsErrorShown] = useState<boolean>(false);
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+  useEffect(() => {
```

Comment:
I'd love a line break between the variable declarations block and the `useEffect()` call to improve readability.